src/Tests/VCEL.Test/FunctionTests.cs

```diff
@@ -1,384 +1,515 @@
+    [Test]
+    public void EvalNow()
+    {
+        var exprString = "now()";
+        foreach (var parseResult in CompositeExpression.ParseMultiple(exprString))
         {
-            var exprString = "now()";
-            foreach (var parseResult in CompositeExpression.ParseMultiple(exprString))
-            {
-                var expr = parseResult.Expression;
-                var value = expr.Evaluate(new { });
-                Assert.That(value, Is.EqualTo(DateTime.Now).Within(TimeSpan.FromSeconds(1)));
-            }
+            var expr = parseResult.Expression;
+            var value = expr.Evaluate(new { });
+            Assert.That(value, Is.EqualTo(DateTime.Now).Within(TimeSpan.FromSeconds(1)));
         }
+    }
 
-        [Test]
-        public void EvalToday()
+    [TestCase("today()", 0)]
+    [TestCase("today(1)", 1)]
+    [TestCase("today(-1)", -1)]
+    [TestCase("today(100)", 100)]
+    public void EvalToday(string exprString, int offset)
+    {
+        foreach (var parseResult in CompositeExpression.ParseMultiple(exprString))
         {
-            var exprString = "today()";
-            foreach (var parseResult in CompositeExpression.ParseMultiple(exprString))
-            {
-                var expr = parseResult.Expression;
-                var value = expr.Evaluate(new { });
-                Assert.That(value, Is.EqualTo(DateTime.Today));
-            }
+            var expr = parseResult.Expression;
+            var value = expr.Evaluate(new { });
+            Assert.That(value, Is.EqualTo(DateTime.Today.AddDays(offset)));
         }
+    }
 
-        [TestCase("lowercase('ABCD')", "abcd")]
-        [TestCase("uppercase('abcd')", "ABCD")]
-        [TestCase("substring('test_substring', 5)", "substring")]
-        [TestCase("substring('test_substring', 0, 4)", "test")]
-        [TestCase("split('test_split', '_')", new[] { "test", "split" })]
-        [TestCase("replace('test_replace', '_replace', '')", "test")]
-        [TestCase("replace('test_replace', '_replace', '_test')", "test_test")]
-        public void EvalStringFunctions(string exprString, object expected)
+    [TestCase("lowercase('ABCD')", "abcd")]
+    [TestCase("lowercase(null)", null)]
+    [TestCase("uppercase('abcd')", "ABCD")]
+    [TestCase("uppercase(null)", null)]
+    [TestCase("substring('test_substring', 5)", "substring")]
+    [TestCase("substring('test_substring', -5)", null)]
+    [TestCase("substring('test_substring', 0, 4)", "test")]
+    [TestCase("substring('test_substring', 0, 0)", "")]
+    [TestCase("substring('test_substring', 0, 100)", null)]
+    [TestCase("substring('test_substring', 100, 100)", null)]
+    [TestCase("substring('test_substring', -1, 100)", null)]
+    [TestCase("substring('test_substring', -1, -1)", null)]
+    [TestCase("split('test_split', '_')", new[] { "test", "split" })]
+    [TestCase("split('test_split', null)", new string[0])]
+    [TestCase("split(null, '_')", new string[0])]
```

Comment:
I would expect these two cases to return null

Reply:
Updated